File: src/secret_store.rs
use std::pin::Pin;

use cxx::{CxxString, CxxVector};

use crate::{
    error::{ErrPtr, FastlyStatusWrapper},
    try_fe,
};

pub struct Secret(pub(crate) fastly::secret_store::Secret);

pub fn m_static_secret_store_secret_from_bytes(
    bytes: &CxxVector<u8>,
    mut out: Pin<&mut *mut Secret>,
    mut err: ErrPtr,
) {
    out.set(Box::into_raw(Box::new(Secret(try_fe!(
        err,
        fastly::secret_store::Secret::from_bytes(bytes.as_slice().to_owned())
            .map_err(FastlyStatusWrapper)
    )))))
}

impl Secret {
    pub fn plaintext(&self, out: Pin<&mut CxxString>) {
        out.push_bytes(self.0.plaintext().as_ref());
    }
}

pub struct SecretStore(pub(crate) fastly::SecretStore);

pub fn m_static_secret_store_secret_store_open(
    name: &CxxString,
    mut out: Pin<&mut *mut SecretStore>,
    mut err: ErrPtr,
) {
    out.set(Box::into_raw(Box::new(SecretStore(try_fe!(
        err,
        fastly::SecretStore::open(try_fe!(err, name.to_str()))
    )))))
}

impl SecretStore {
    pub fn get(&self, key: &CxxString, mut out: Pin<&mut *mut Secret>, mut err: ErrPtr) {
        out.set(
            self.0
                .get(try_fe!(err, key.to_str()))
                .map(Secret)
                .map(Box::new)
                .map(Box::into_raw)
                .unwrap_or_else(std::ptr::null_mut),
        );
    }

    pub fn contains(&self, key: &CxxString, mut err: ErrPtr) -> bool {
        try_fe!(err, self.0.contains(try_fe!(err, key.to_str())))
    }
}

File: src/cpp/secret_store.cpp
#include "util.h"
#include <fastly/error.h>
#include <fastly/sdk-sys.h>
#include <fastly/secret_store.h>

namespace fastly::secret_store {

fastly::expected<Secret> Secret::from_bytes(std::vector<uint8_t> data) {
  fastly::sys::secret_store::Secret *out;
  fastly::sys::error::FastlyError *err;
  fastly::sys::secret_store::m_static_secret_store_secret_from_bytes(data, out,
                                                                     err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return FSLY_BOX(secret_store, Secret, out);
  }
}

std::string Secret::plaintext() {
  std::string out;
  this->s->plaintext(out);
  return out;
}

fastly::expected<SecretStore> SecretStore::open(std::string_view name) {
  fastly::sys::secret_store::SecretStore *out;
  fastly::sys::error::FastlyError *err;
  fastly::sys::secret_store::m_static_secret_store_secret_store_open(
      static_cast<std::string>(name), out, err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return FSLY_BOX(secret_store, SecretStore, out);
  }
}

fastly::expected<std::optional<Secret>> SecretStore::get(std::string_view key) {
  fastly::sys::secret_store::Secret *out;
  fastly::sys::error::FastlyError *err;
  this->ss->get(static_cast<std::string>(key), out, err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else if (out != nullptr) {
    return std::optional<Secret>(FSLY_BOX(secret_store, Secret, out));
  } else {
    return std::nullopt;
  }
}

fastly::expected<bool> SecretStore::contains(std::string_view key) {
  fastly::sys::error::FastlyError *err;
  bool out{this->ss->contains(static_cast<std::string>(key), err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<bool>(out);
  }
}

} // namespace fastly::secret_store
